src/samples/audio/smart_amp_test.c

```diff
@@ -86,6 +105,213 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
+		comp_err(dev, "smart_amp_set_config(): invalid config size %u, expect %u",
+			 cfg_size, sizeof(struct sof_smart_amp_config));
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "smart_amp_set_config(): config size = %u", cfg_size);
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg,
+		 sizeof(struct sof_smart_amp_config));
+
+	return 0;
+}
+
+static inline int smart_amp_get_config(struct comp_dev *dev, char *data,
+				       uint32_t *data_size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	uint32_t cfg_size;
+	int ret;
+
+	cfg_size = sizeof(struct smart_amp_data);
+
+	if (cfg_size > *data_size) {
+		comp_err(dev, "smart_amp_get_config(): wrong config size %d",
+			 *data_size);
+	}
+
+	*data_size = cfg_size;
+
+	ret = memcpy_s(data, cfg_size, &sad->config, cfg_size);
+	assert(!ret);
+
+	return 0;
+}
+
+static int smart_amp_set_large_config(struct comp_dev *dev,
+				      uint32_t param_id,
+				      bool first_block,
+				      bool last_block,
+				      uint32_t data_offset,
+				      const char *data)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "smart_amp_set_large_config()");
+
+	switch (param_id) {
+	case SMART_AMP_SET_MODEL:
+		return ipc4_comp_data_blob_set(sad->model_handler,
+					       first_block,
+					       last_block,
+					       data_offset,
+					       data);
+	case SMART_AMP_SET_CONFIG:
+		return smart_amp_set_config(dev, data, data_offset);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int smart_amp_get_large_config(struct comp_dev *dev,
+				      uint32_t param_id,
+				      bool first_block,
+				      bool last_block,
+				      uint32_t *data_offset,
+				      char *data)
+{
+	comp_dbg(dev, "smart_amp_get_large_config()");
+
+	switch (param_id) {
+	case SMART_AMP_GET_CONFIG:
+		return smart_amp_get_config(dev, data, data_offset);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int smart_amp_get_attribute(struct comp_dev *dev,
+				   uint32_t type,
+				   void *value)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "smart_amp_get_attribute()");
+
+	switch (type) {
+	case COMP_ATTR_BASE_CONFIG:
+		*(struct ipc4_base_module_cfg *)value = sad->config.base;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int smart_amp_bind(struct comp_dev *dev, void *data)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *buffer_c;
+	struct comp_buffer *source_buffer;
+	struct list_item *blist;
+
+	comp_dbg(dev, "smart_amp_bind()");
+
+	/* searching for feedback source buffers */
+	list_for_item(blist, &dev->bsource_list) {
+		source_buffer = container_of(blist, struct comp_buffer, sink_list);
+
+		k_mutex_lock(&sad->lock, K_FOREVER);
+		buffer_c = buffer_acquire(source_buffer);
+		if (IPC4_SINK_QUEUE_ID(buffer_c->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID) {
+			sad->feedback_buf = source_buffer;
+			buffer_c->stream.channels = sad->config.feedback_channels;
```

Comment:
@RanderWang Im not sure where my comment went but I was wondering can be done in prepare instead of bind?

Reply:
it is at https://github.com/thesofproject/sof/pull/6751#discussion_r1049219248